src/Testy.jl

```diff
@@ -3,38 +3,73 @@ module Testy
 using Test
 
 # TODO refactor into patch to bring these into Base
-"""
-Helper for `pmatch`.
-"""
-function pexec(re,subject,offset,options,match_data)
+
+# NOTE: Since  the internals have changed in julia v1.3.0, we need VERSION checks inside
+# these functions.
+
+# Helper for `pmatch`: Mirrors Base.PCRE.exec
+function pexec(re, subject, offset, options, match_data)
     rc = ccall((:pcre2_match_8, Base.PCRE.PCRE_LIB), Cint,
                (Ptr{Cvoid}, Ptr{UInt8}, Csize_t, Csize_t, Cuint, Ptr{Cvoid}, Ptr{Cvoid}),
-               re, subject, sizeof(subject), offset, options, match_data, Base.PCRE.MATCH_CONTEXT[])
+               re, subject, sizeof(subject), offset, options, match_data,
+               @static if VERSION >= v"1.3-"
+                   Base.PCRE.get_local_match_context()
+               else
+                   Base.PCRE.MATCH_CONTEXT[]
+               end)
     # rc == -1 means no match, -2 means partial match.
     rc < -2 && error("PCRE.exec error: $(err_message(rc))")
     rc
 end
 
+# Helper to call `pexec` w/ thread-safe match data: Mirrors Base.PCRE.exec_r_data
+# Only used for Julia Version >= v"1.3-"
+function pexec_r_data(re, subject, offset, options)
+    match_data = Base.PCRE.create_match_data(re)
+    ans = pexec(re, subject, offset, options, match_data)
+    return ans, match_data
+end
+
 """
+   pmatch(r::Regex, s::AbstractString[, idx::Integer[, addopts]])
+
 Variant of `Base.match` that supports partial matches (when `Base.PCRE.PARTIAL_HARD`
 is set in `re.match_options`).
 """
 function pmatch(re::Regex, str::Union{SubString{String}, String}, idx::Integer, add_opts::UInt32=UInt32(0))
     Base.compile(re)
     opts = re.match_options | add_opts
-    # rc == -1 means no match, -2 means partial match.
-    rc = pexec(re.regex, str, idx-1, opts, re.match_data)
-    if rc == -1 || (rc == -2 && (re.match_options & Base.PCRE.PARTIAL_HARD) == 0)
-        return nothing
+    @static if VERSION >= v"1.3-"
+        # rc == -1 means no match, -2 means partial match.
+        rc, data = pexec_r_data(re.regex, str, idx-1, opts)
+        if rc == -1 || (rc == -2 && (re.match_options & Base.PCRE.PARTIAL_HARD) == 0)
```

Comment:
@tjgreen Shouldn't this be checking `opts` instead of `re.match_options` for `PARTIAL_HARD`? If the user adds `PARTIAL_HARD` in the `add_opts`, it looks like that gets passed-through and should be used in this return value?